command/job_restart.go

```diff
@@ -0,0 +1,850 @@
+	if err != nil {
+		c.ui.Error(err.Error())
+		c.ui.Error(commandErrorText(c))
+		return code
+	}
+	if code != 0 {
+		return code
+	}
+
+	c.client, err = c.Meta.Client()
+	if err != nil {
+		c.ui.Error(fmt.Sprintf("Error initializing client: %v", err))
+		return 1
+	}
+
+	// Check if the job exists.
+	// Avoid prefix matching to make sure we only restart the expected job.
+	job, _, err := c.client.Jobs().Info(c.jobID, nil)
+	if err != nil {
+		c.ui.Error(fmt.Sprintf("Error retrieving job %q: %s", c.jobID, err))
+		return 1
+	}
+
+	// Check if passed groups and tasks are valid.
+	err = c.validateGroupsAndTasks(job)
+	if err != nil {
+		c.ui.Error(err.Error())
+		return 1
+	}
+
+	// Fetch all allocations for the job and filter out the ones that are not
+	// eligible for restart.
+	allocs, _, err := c.client.Jobs().Allocations(c.jobID, false, nil)
+	if err != nil {
+		c.ui.Error(fmt.Sprintf("Error retrieving allocations for jobs %s: %v", c.jobID, err))
+		return 1
+	}
+
+	restartAllocs := make([]string, 0, len(allocs))
+	for _, alloc := range allocs {
+		shortAllocID := limit(alloc.ID, c.length)
+
+		// Skip allocations that are not running.
+		allocRunning := alloc.ClientStatus == api.AllocClientStatusRunning || alloc.DesiredStatus == api.AllocDesiredStatusRun
+		if !allocRunning {
+			if c.verbose {
+				c.ui.Output(c.Colorize().Color(fmt.Sprintf(
+					"[dark_gray]    %s: Skipping allocation %s because desired status is %q and client status is %q[reset]",
+					formatTime(time.Now()),
+					shortAllocID,
+					alloc.ClientStatus,
+					alloc.DesiredStatus,
+				)))
+			}
+			continue
+		}
+
+		// Skip allocations for groups that were not requested.
+		if c.groups.Size() > 0 && !c.groups.Contains(alloc.TaskGroup) {
+			if c.verbose {
+				c.ui.Output(c.Colorize().Color(fmt.Sprintf(
+					"[dark_gray]    %s: Skipping allocation %s because it doesn't have any of requested groups[reset]",
+					formatTime(time.Now()),
+					shortAllocID,
+				)))
+			}
+			continue
+		}
+
+		// Skip allocations that don't have any of the requested tasks.
+		if c.tasks.Size() > 0 {
+			hasTask := false
+			for task := range alloc.TaskStates {
+				if c.tasks.Contains(task) {
+					hasTask = true
+					break
+				}
+			}
+			if !hasTask {
+				if c.verbose {
+					c.ui.Output(c.Colorize().Color(fmt.Sprintf(
+						"[dark_gray]    %s: Skipping allocation %s because it doesn't have any of requested tasks[reset]",
+						formatTime(time.Now()),
+						shortAllocID,
+					)))
+				}
+				continue
+			}
+		}
+
+		restartAllocs = append(restartAllocs, alloc.ID)
+	}
+
+	// Exit early if there's nothing to do.
+	if len(restartAllocs) == 0 {
+		c.ui.Output("No allocations to restart")
+		return 0
+	}
+
+	// Calculate absolute batch size based on the number of eligible
+	// allocations if the value provided is a percentage.
+	// Round values up to increase parallelism.
+	if c.batchSizePercent {
+		c.batchSize = int(math.Ceil(float64(len(restartAllocs)*c.batchSize) / 100))
+	}
```

Comment:
Should we worry about large absolute batch sizes? I first thought of a percent batch size as a convenience option, but a `nomad job restart -reschedule -batch-size=100%` in a job with hundreds of allocations could cause havoc in a cluster 😅

Reply:
`Alloc.Stop`, which is used when `-reschedule` is set, [does create a Raft entry](https://github.com/hashicorp/nomad/blob/v1.4.4/nomad/alloc_endpoint.go#L310-L325) for each allocation, but the FSM operation [already supports batching multiple allocations in the same Raft transaction](https://github.com/hashicorp/nomad/blob/main/nomad/state/state_store.go#L3888-L3893), so that's something we can improve in the future.